examples/ppo/main.py

```diff
@@ -0,0 +1,309 @@
+from typing import Tuple, List
+from queue import Queue
+from absl import flags
+from absl import app
+import threading
+
+
+from models import create_model, create_optimizer
+from agent import policy_action
+from remote import RemoteSimulator
+from test_episodes import test
+from env import get_num_actions
+
+FLAGS = flags.FLAGS
+
+# default hyperparameters taken from PPO paper and openAI baselines 2
+# https://github.com/openai/baselines/blob/master/baselines/ppo2/defaults.py
+
+flags.DEFINE_float(
+    'learning_rate', default=2.5e-4,
+    help=('The learning rate for the Adam optimizer.')
+)
+
+flags.DEFINE_integer(
+    'batch_size', default=256,
+    help=('Batch size for training.')
+)
+
+flags.DEFINE_integer(
+    'num_agents', default=8,
+    help=('Number of agents playing in parallel.')
+)
+
+flags.DEFINE_integer(
+    'actor_steps', default=128,
+    help=('Batch size for training.')
+)
+
+flags.DEFINE_integer(
+    'num_epochs', default=3,
+    help=('Number of epochs per each unroll of the policy.')
+)
+
+flags.DEFINE_float(
+    'gamma', default=0.99,
+    help=('Discount parameter.')
+)
+
+flags.DEFINE_float(
+    'lambda_', default=0.95,
+    help=('Generalized Advantage Estimation parameter.')
+)
+
+flags.DEFINE_float(
+    'clip_param', default=0.1,
+    help=('The PPO clipping parameter used to clamp ratios in loss function.')
+)
+
+flags.DEFINE_float(
+    'vf_coeff', default=0.5,
+    help=('Weighs value function loss in the total loss.')
+)
+
+flags.DEFINE_float(
+    'entropy_coeff', default=0.01,
+    help=('Weighs entropy bonus in the total loss.')
+)
+
+@partial(jax.vmap, in_axes=(1, 1, 1, None, None), out_axes=1)
+@jax.jit
+def gae_advantages(rewards, terminal_masks, values, discount, gae_param):
+  """Use Generalized Advantage Estimation (GAE) to compute advantages
+  Eqs. (11-12) in PPO paper arXiv: 1707.06347.
+  Uses key observation that A_{t} = \delta_t + \gamma*\lambda*A_{t+1}.
+  """
+  assert rewards.shape[0] + 1 == values.shape[0], ("One more value needed; "
+          "Eq. (12) in PPO paper requires V(s_{t+1}) to calculate \delta_t")
+  advantages, gae = [], 0.
+  for t in reversed(range(len(rewards))):
+    # masks to set next state value to 0 for terminal states
+    value_diff = discount * values[t + 1] * terminal_masks[t] - values[t]
+    delta = rewards[t] + value_diff
+    # masks[t] to ensure that values before and after a terminal state
+    # are independent of each other
+    gae = delta + discount * gae_param * terminal_masks[t] * gae
+    advantages = [gae] + advantages
+  return jnp.array(advantages)
+
+@jax.jit
+def train_step(optimizer, trn_data, clip_param, vf_coeff, entropy_coeff):
+  """Compilable train step. Runs an entire epoch of training (i.e. the loop over
+  minibatches within an epoch is included here for performance reasons.
+  Args:
+    optimizer: optimizer for the actor-critic model
+    trn_data: Tuple of the following five elements forming the experience:
```

Comment:
Since https://github.com/google/flax/pull/470/commits/7b02ec079fce275d92302adf618775bea7ba564d, `trn_data` has been renamed to `trajectories`, thanks for the suggestion!